### PR TITLE
Add pre-generated bindings

### LINKED ISSUE #15: How to make the build process more robust (and more portable)
This issue is mainly motivated by the MacOS-specific fixes in #14, but also by the fact that we currently use 4 different environment variables to control the build process for this package (6 variables if you count convenience variables as well).

So far, the installation process for P4est.jl can be described roughly as follows:
1. We first find out where the p4est library and header files are located.
  * If this is a p4est-build with MPI support, we also figure out where the MPI header files are located.
2. We then generate C bindings for p4est using [CBindingGen.jl](https://github.com/analytech-solutions/CBindingGen.jl) and save them to a local file
3. In the P4est.jl package itself, we include the file with the bindings and add additional `using`s etc.

This procedure follows the recommendations by CBindingGen.jl, which states in its [README.md](https://github.com/analytech-solutions/CBindingGen.jl#usage):
> This package should only be used at package build time when you wish to generate bindings to a particular C library on the system or one built and installed at build time. The generated bindings file can then be included from your package code. Bindings files created by this package should not be committed with your package and they are not meant for editing by lowly humans once generated.

However, there at least two issues with this approach:
* **It requires users to *have* all header files**: Since we need the header files for the automatic generation of the C bindings, they must be available on the user's machine. While this is a reasonable expectation in general for p4est and MPI, it might not be the case if users install them via their package managers. Also, as we can see in #14, it requires to jump through hoops at least on MacOS (and I can't even tell what's going to happen on Windows yet).
* **It requires users to *know* where the header files are located**: This might be again a reasonable expectation on HPC systems, but it's a completely different story for libraries installed via package managers (e.g., on Ubuntu even MPICH and OpenMPI do not have similar include paths).

Both issues could be avoided by using a pre-generated C binding file (or two, for p4est with and without MPI enabled). However, this would probably create new issues, such as dependencies on the versions of p4est and `CBinding.jl`. OTOH, in this case using an external p4est library would only require to specify the path to the p4est shared library file.

A third option could be to combine the current approach with a pre-generated C binding file. In this case I would probably use the pre-generated C bindings file as the default, but allow the user to specify include paths for p4est and MPI and then re-generate the C bindings. The downside of this approach is that it requires an even more complicated build process (that becomes harder and harder to test in the CI), but it would (might?) provide maximum convenience for "standard" users together with full flexibility for advanced users.

To me, this issue blocks #4, so I'd like to solicit your input on this. Personally, I am leaning towards the third option.

### PR DESCRIPTION
**Pull request by sloede**
This allows to use a pre-generated bindings file by default, making it potentially easier to use P4est.jl on other platforms with missing header files.

Closes #15.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 language: julia
 
 julia:
-  - 1 # latest 1.x.y
+  - 1.6 # latest 1.x.y
```
Comment (ranocha) [inline .travis.yml:6]: Why did you change this?

File: Project.toml [hunk 2]
```diff
@@ -12,5 +12,6 @@ Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
 [compat]
 CBinding = "0.9.2"
 MPI = "0.15, 0.16"
+P4est_jll = "~2.3.1"
```
Comment (ranocha) [inline Project.toml:15]:
```suggestion
P4est_jll = "= 2.3.1"
```
Do you really want a tilde specifier? If you want to pick a specific version, you need the [equality specifier](https://pkgdocs.julialang.org/v1/compatibility/#Equality-specifier).

File: deps/Project.toml [hunk 3]
```diff
@@ -8,3 +8,4 @@ Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
 [compat]
 CBindingGen = "0.4"
 MPI = "0.15, 0.16"
+P4est_jll = "~2.3.1"
```
Comment (ranocha) [inline deps/Project.toml:11]: See above (feel free to resolve at the same time)

File: deps/build.jl [hunk 4]
```diff
@@ -4,6 +4,8 @@ using MPI
 import Pkg.TOML
 import P4est_jll
 
+const pre_generated_bindings_filename = joinpath(@__DIR__, "libp4est-2.3.1.jl")
```
Comment (ranocha) [inline deps/build.jl:7]: This hard-codes the exact version number (equality vs. tilde specifier above)

**Review by ranocha — CHANGES_REQUESTED** [review]
Thanks for working on this!
- Most changes in `deps/build.jl` were just indentation changes, correct? (I didn't read all of them carefully)
- Do we really want to push 17 LOC auto-generated stuff here?

File: README.md [hunk 5]
```diff
@@ -15,7 +15,7 @@ connected adaptive quadtrees or octrees in parallel.
 ## Installation
 If you have not yet installed Julia, please [follow the instructions for your
 operating system](https://julialang.org/downloads/platform/). P4est.jl works
-with Julia v1.5.
+with Julia v1.6.
```
Comment (ranocha) [inline README.md:18]: If we require this, we should also change `Project.toml` accordingly. If not, we should probably also add tests on Julia v1.5.

File: .travis.yml (hunk 1 above)
Comment (sloede) [reply .travis.yml:6]: Because `P4est_jll.jl@2.3.1` [requires Julia 1.6](https://github.com/JuliaBinaryWrappers/P4est_jll.jl/blob/c281fee10ffd8f172859055b4ea9c6b5f26bc077/Project.toml#L14) and this seemed to be the underlying cause for troubles once I started pinning the JLL version to v2.3.1.

File: Project.toml (hunk 2 above)
Comment (sloede) [reply Project.toml:15]: I assume that if they follow semver at p4est, all patch releases should still work, but since `2.3.0` was not officially released I cannot vouch for it, thus the tilde.

File: deps/build.jl (hunk 4 above)
Comment (sloede) [reply deps/build.jl:7]: Yes, I thought this would make it easier in the future to understand from which exact version these bindings have been generated, in case a subsequent release of p4est breaks stuff.

File: README.md (hunk 5 above)
Comment (sloede) [reply README.md:18]: Good point. Fixed in 1b9a2b1.

File: Project.toml (hunk 2 above)
Comment (sloede) [reply Project.toml:15]: Well, promise is maybe too strong, but I think it's a reasonable assumption

**Review by ranocha — APPROVED** [review]
Maybe it's okay to gather the auto generated bindings here.

**Comment by sloede** [comment]
> Maybe it's okay to gather the auto generated bindings here.

I don't like it... I already feel that I'll likely need to experiment with the bindings, especially with switching to `CBinding.jl@1`, and I know I'll hate it if I have to always wait and consider if it is ok to push it to `main` or not. I think I'll try out to put it in a gist as an artifact for now before merging this, and we can always change it later.

**Comment by sloede** [comment]
OK, this is weird. Apparently you need have the `Artifacts.toml` both in the root and in the `deps` directory to make it work. I feel like this is not a very good solution either :-/

**Comment by sloede** [comment]
This is not yet optimal, with the redundant `Artifacts.toml`. However, I'd like to move forward to allow @efaulhaber to proceed with his project as well.

**Review by ranocha — APPROVED** [review]
